__The following code will compare two potential portfolios, containing ten stocks each. One contains stocks from only the 
consumer staples sector, while the other contains stocks from ten different industries. We will examine the expected return 
for each portfolio using the efficient frontier method. The objective is to determine whether the diverse portfolio
provides a better anticipated return than the portfolio that draws from only one sector.__

In [31]:
import pandas as pd
from pandas_datareader import data as pdr
import datetime as dt

In [7]:
pd.set_option=('display.max_columns', None)
pd.set_option=('display.max_rows', None)

__Choose date range to measure historical data.__

In [8]:
start = dt.datetime(2000,1,1)
end=dt.datetime.now()

__Write UDF to import stock data.__

In [9]:
def get_stock(ticker):
    data = pdr.DataReader(f'{ticker}', 'yahoo', start, end)
    data[f'{ticker}'] = data['Adj Close']
    data = data[[f'{ticker}']]
    print(data.head())
    return data

__Use the function to read data for the following stocks from the consumer staples sector:__

In [10]:
coke = get_stock('KO')
keurig = get_stock('KDP')
costco = get_stock('COST')
target = get_stock('TGT')
unilever = get_stock('UL')
procter = get_stock('PG')
genmills = get_stock('GIS')
mondelez = get_stock('MDLZ')
ulta = get_stock('ULTA')
revlon = get_stock('REV')

                   KO
Date                 
2000-01-03  15.452650
2000-01-04  15.469786
2000-01-05  15.606835
2000-01-06  15.623975
2000-01-07  16.651869
                 KDP
Date                
2008-05-07  3.019603
2008-05-08  3.007763
2008-05-09  2.988816
2008-05-12  3.027893
2008-05-13  3.066971
                 COST
Date                 
2000-01-03  31.135601
2000-01-04  29.430117
2000-01-05  29.933004
2000-01-06  30.534292
2000-01-07  32.556789
                  TGT
Date                 
2000-01-03  23.972727
2000-01-04  22.933144
2000-01-05  22.413357
2000-01-06  21.352989
2000-01-07  22.454939
                  UL
Date                
2000-01-03  7.892837
2000-01-04  7.777974
2000-01-05  8.106159
2000-01-06  8.270248
2000-01-07  8.729709
                   PG
Date                 
2000-01-03  29.713684
2000-01-04  29.141945
2000-01-05  28.587513
2000-01-06  29.904293
2000-01-07  32.295235
                 GIS
Date                
2000-01-03  8.847642
2000-01-04  8.687069
2000-0

In [11]:
from functools import reduce

__Write UDF to combine info for all stocks into one variable.__

In [18]:
def combine_stocks(tickers):
    data_frames = []
    for i in tickers:
        data_frames.append(get_stock(i))
    df_merged = reduce(lambda left, right: pd.merge(left, right, on=['Date'], how='outer'), data_frames)
    print(df_merged.head())
    return df_merged

__Use the above function to combine stock data.__

In [19]:
staples_stocks = ['KO', 'KDP', 'COST', 'TGT', 'UL', 'PG', 'GIS', 'MDLZ', 'ULTA', 'REV']
staples_portfolio = combine_stocks(staples_stocks)

                   KO
Date                 
2000-01-03  15.452653
2000-01-04  15.469783
2000-01-05  15.606839
2000-01-06  15.623977
2000-01-07  16.651865
                 KDP
Date                
2008-05-07  3.019604
2008-05-08  3.007764
2008-05-09  2.988816
2008-05-12  3.027893
2008-05-13  3.066971
                 COST
Date                 
2000-01-03  31.135590
2000-01-04  29.430122
2000-01-05  29.933012
2000-01-06  30.534290
2000-01-07  32.556789
                  TGT
Date                 
2000-01-03  23.972729
2000-01-04  22.933136
2000-01-05  22.413353
2000-01-06  21.352987
2000-01-07  22.454945
                  UL
Date                
2000-01-03  7.892838
2000-01-04  7.777972
2000-01-05  8.106162
2000-01-06  8.270250
2000-01-07  8.729709
                   PG
Date                 
2000-01-03  29.713697
2000-01-04  29.141937
2000-01-05  28.587511
2000-01-06  29.904284
2000-01-07  32.295231
                 GIS
Date                
2000-01-03  8.847645
2000-01-04  8.687068
2000-0

__Print a sample of the portfolio.__

In [20]:
staples_portfolio

KO        KDP        COST         TGT         UL  \
Date                                                                  
2000-01-03  15.452653        NaN   31.135590   23.972729   7.892838   
2000-01-04  15.469783        NaN   29.430122   22.933136   7.777972   
2000-01-05  15.606839        NaN   29.933012   22.413353   8.106162   
2000-01-06  15.623977        NaN   30.534290   21.352987   8.270250   
2000-01-07  16.651865        NaN   32.556789   22.454945   8.729709   
...               ...        ...         ...         ...        ...   
2022-03-07  61.080002  38.139999  528.520020  211.039993  43.880001   
2022-03-07  61.080002  38.139999  528.520020  211.039993  43.880001   
2022-03-07  61.080002  38.139999  528.520020  211.039993  43.880001   
2022-03-07  61.080002  38.139999  528.520020  211.039993  43.880001   
2022-03-07  61.080002  38.139999  528.520020  211.039993  43.880001   

                    PG        GIS       MDLZ        ULTA        REV  
Date                                                                 
2000-01-03   29.713697   8.847645        NaN         NaN  71.260963  
2000-01-04   29.141937   8.687068        NaN         NaN  75.349709  
2000-01-05   28.587511   8.638897        NaN         NaN  70.676857  
2000-01-06   29.904284   8.693941        NaN         NaN  71.845070  
2000-01-07   32.295231   8.710131        NaN         NaN  71.260963  
...                ...        ...        ...         ...        ...  
2022-03-07  152.839996  67.669998  62.529999  350.839996   7.740000  
2022-03-07  152.839996  67.669998  62.529999  350.839996   7.740000  
2022-03-07  152.839996  67.669998  62.529999  350.839996   7.740000  
2022-03-07  152.839996  67.669998  62.529999  350.839996   7.740000  
2022-03-07  152.839996  67.669998  62.529999  350.839996   7.740000  

[6603 rows x 10 columns]

In [21]:
pip install PyPortfolioOpt

Note: you may need to restart the kernel to use updated packages.


__Find mean historical return and covariance for the portfolio.__

In [22]:
from pypfopt.expected_returns import mean_historical_return
from pypfopt.risk_models import CovarianceShrinkage

mu_staples = mean_historical_return(staples_portfolio)
s_staples = CovarianceShrinkage(staples_portfolio).ledoit_wolf()

__Use the Efficient Frontier method to evaluate the portfolio.__

In [23]:
from pypfopt.efficient_frontier import EfficientFrontier

In [24]:
staples_ef = EfficientFrontier(mu_staples, s_staples)
staples_weights = staples_ef.max_sharpe()

__Find optimal weights for each stock in the portfolio.__

In [25]:
staples_cleaned_wts = staples_ef.clean_weights()
print(dict(staples_cleaned_wts))

{'KO': 0.0, 'KDP': 0.62627, 'COST': 0.13853, 'TGT': 0.0, 'UL': 0.0, 'PG': 0.0, 'GIS': 0.15257, 'MDLZ': 0.0, 'ULTA': 0.08264, 'REV': 0.0}


__Find the expected return of the portfolio.__

In [26]:
staples_ef.portfolio_performance(verbose=True)

Expected annual return: 13.5%
Annual volatility: 15.2%
Sharpe Ratio: 0.76


(0.1354909816501993, 0.15235780955012299, 0.7580246919486252)

__As we can see above, the expected return is acceptable (>10%) and the volatility is average (approx. 15%). However, the Sharpe ratio is low, which is cause for concern with this mix of stocks.__

__Next, determine how to allocate our capital among the stocks. We have $10,000 to invest.__

In [27]:
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices

In [28]:
staples_prices = get_latest_prices(staples_portfolio)

In [29]:
staples_da = DiscreteAllocation(staples_weights, staples_prices, total_portfolio_value = 10000)
staples_allocation, staples_leftover = staples_da.greedy_portfolio()

In [30]:
print("Discrete Allocation: ", staples_allocation)
print("Funds remaining: ${:.2f}".format(staples_leftover))

Discrete Allocation:  {'KDP': 164, 'GIS': 23, 'COST': 2, 'ULTA': 3}
Funds remaining: $79.07


__Repeat using a combination of stocks from different sectors.__

__Use the get_stock function to import data.__

In [32]:
disney = get_stock('DIS')
valero = get_stock('VLO')
ibm = get_stock('IBM')
pfizer = get_stock('PFE')
usbank = get_stock('USB')
_3M = get_stock('MMM')
target = get_stock('TGT')
mcdonalds = get_stock('MCD')
duke = get_stock('DUK')
dupont = get_stock('DD')

                  DIS
Date                 
2000-01-03  23.115252
2000-01-04  24.469284
2000-01-05  25.484808
2000-01-06  24.469284
2000-01-07  24.082417
                 VLO
Date                
2000-01-03  2.610008
2000-01-04  2.584420
2000-01-05  2.678243
2000-01-06  2.763538
2000-01-07  2.729421
                  IBM
Date                 
2000-01-03  66.433662
2000-01-04  64.178673
2000-01-05  66.433662
2000-01-06  65.288284
2000-01-07  65.001930
                  PFE
Date                 
2000-01-03  14.201269
2000-01-04  13.672202
2000-01-05  13.894965
2000-01-06  14.396190
2000-01-07  15.370790
                  USB
Date                 
2000-01-03  10.365388
2000-01-04   9.987318
2000-01-05   9.766777
2000-01-06  10.522913
2000-01-07  10.932488
                  MMM
Date                 
2000-01-03  26.701992
2000-01-04  25.640985
2000-01-05  26.383694
2000-01-06  28.505699
2000-01-07  29.071579
                  TGT
Date                 
2000-01-03  23.972723
2000-01-04  22.93

__Combine the individual stock data into one variable.__

In [34]:
diverse_stocks = ['DIS', 'VLO', 'IBM', 'PFE', 'USB', 'MMM', 'TGT', 'MCD', 'DUK', 'DD']
diverse_portfolio = combine_stocks(diverse_stocks)

                  DIS
Date                 
2000-01-03  23.115250
2000-01-04  24.469288
2000-01-05  25.484806
2000-01-06  24.469288
2000-01-07  24.082418
                 VLO
Date                
2000-01-03  2.610008
2000-01-04  2.584419
2000-01-05  2.678243
2000-01-06  2.763538
2000-01-07  2.729419
                  IBM
Date                 
2000-01-03  66.433685
2000-01-04  64.178642
2000-01-05  66.433685
2000-01-06  65.288261
2000-01-07  65.001923
                  PFE
Date                 
2000-01-03  14.201269
2000-01-04  13.672200
2000-01-05  13.894964
2000-01-06  14.396184
2000-01-07  15.370778
                  USB
Date                 
2000-01-03  10.365387
2000-01-04   9.987318
2000-01-05   9.766777
2000-01-06  10.522915
2000-01-07  10.932488
                  MMM
Date                 
2000-01-03  26.701996
2000-01-04  25.640995
2000-01-05  26.383698
2000-01-06  28.505701
2000-01-07  29.071581
                  TGT
Date                 
2000-01-03  23.972729
2000-01-04  22.93

__View a sample to verify data is correct.__

In [35]:
diverse_portfolio

DIS        VLO         IBM        PFE        USB  \
Date                                                                  
2000-01-03   23.115250   2.610008   66.433685  14.201269  10.365387   
2000-01-04   24.469288   2.584419   64.178642  13.672200   9.987318   
2000-01-05   25.484806   2.678243   66.433685  13.894964   9.766777   
2000-01-06   24.469288   2.763538   65.288261  14.396184  10.522915   
2000-01-07   24.082418   2.729419   65.001923  15.370778  10.932488   
...                ...        ...         ...        ...        ...   
2022-03-07  133.500000  84.010002  126.070000  47.980000  53.099998   
2022-03-07  133.500000  84.010002  126.070000  47.980000  53.099998   
2022-03-07  133.500000  84.010002  126.070000  47.980000  53.099998   
2022-03-07  133.500000  84.010002  126.070000  47.980000  53.099998   
2022-03-07  133.500000  84.010002  126.070000  47.980000  53.099998   

                   MMM         TGT         MCD         DUK         DD  
Date                                                                   
2000-01-03   26.701996   23.972729   22.930809   14.990426  30.408072  
2000-01-04   25.640995   22.933149   22.460611   15.223432  29.576929  
2000-01-05   26.383698   22.413349   22.822302   15.844806  30.537056  
2000-01-06   28.505701   21.352980   22.496788   16.271988  31.396845  
2000-01-07   29.071581   22.454935   23.075485   16.699179  32.099022  
...                ...         ...         ...         ...        ...  
2022-03-07  143.279999  211.039993  224.330002  106.699997  69.070000  
2022-03-07  143.279999  211.039993  224.330002  106.699997  69.070000  
2022-03-07  143.279999  211.039993  224.330002  106.699997  69.070000  
2022-03-07  143.279999  211.039993  224.330002  106.699997  69.070000  
2022-03-07  143.279999  211.039993  224.330002  106.699997  69.070000  

[6603 rows x 10 columns]

__Find mean historical return and covariance for the portfolio.__

In [36]:
mu_div = mean_historical_return(diverse_portfolio)
s_div = CovarianceShrinkage(diverse_portfolio).ledoit_wolf()

__Use Efficient Frontier method to evaluate portfolio.__

In [37]:
diverse_ef = EfficientFrontier(mu_div, s_div)
diverse_weights = diverse_ef.max_sharpe()

__Find optimal weights for each stock.__

In [38]:
diverse_cleaned_wts = diverse_ef.clean_weights()
print(dict(diverse_cleaned_wts))

{'DIS': 0.0, 'VLO': 0.24381, 'IBM': 0.0, 'PFE': 0.0, 'USB': 0.0, 'MMM': 0.0, 'TGT': 0.10497, 'MCD': 0.42889, 'DUK': 0.22233, 'DD': 0.0}


__Find the expected return of the portfolio.__

In [39]:
diverse_ef.portfolio_performance(verbose=True)

Expected annual return: 10.0%
Annual volatility: 18.8%
Sharpe Ratio: 0.43


(0.09993759547939379, 0.1875312901939722, 0.4262627073951802)

__The expected return is acceptable, although it is lower than the first portfolio. The volatility is above average, at 18.8%, and the Sharpe ratio is low. This portfolio appears to be the worse option of the two available here.__

__Determine how to allocate our funds. We have $10,000 to invest.__

In [40]:
div_prices = get_latest_prices(diverse_portfolio)

In [41]:
diverse_da = DiscreteAllocation(diverse_weights, div_prices, total_portfolio_value = 10000)
diverse_allocation, diverse_leftover = diverse_da.greedy_portfolio()

In [42]:
print("Discrete Allocation: ", diverse_allocation)
print("Funds remaining: ${:.2f}".format(diverse_leftover))

Discrete Allocation:  {'MCD': 19, 'VLO': 29, 'DUK': 21, 'TGT': 5}
Funds remaining: $5.54


__Save both portfolios into CSV files for later use.__

In [43]:
staples_portfolio.to_csv('C:\\Users\\emily\\OneDrive\\Desktop\\staples_portfolio.csv', index=False)
diverse_portfolio.to_csv('C:\\Users\\emily\\OneDrive\\Desktop\\diverse_portfolio.csv', index=False)